### PR TITLE
Vectorize `dotProduct`

### LINKED ISSUE #58096: `dotProduct` is not optimized for vector Similarity
## Background

Vector similarity is very popular right now, and ClickHouse has a few tutorials about this  [Source](https://clickhouse.com/blog/approximate-nearest-neighbour-ann-with-sql-powered-local-sensitive-hashing-lsh-random-projections). Regarding distances between vectors, they mention the most common ones:

- Euclidean distance
- Cosine distance

Captura de pantalla 2023-12-20 a las 21 01 10

But in my opinion dotProduct is a better replacement for cosineDistance because **is the same but faster**, let me explain...

## cosineDistance == dotProduct of L2 normalized vectors

`dotProduct` has a much cheaper computation than `cosineDistance`. But they produce different things.

download

However, **the dotProduct of the L2-normalized vectors** is mathematically the same that cosineDistance, but computationally cheaper because only the norm of the query verctor is computed at search time. The following formula illustrate this:

Captura de pantalla 2023-12-20 a las 21 30 29

>  L2 normalization is dividing (element-wise) the vector by its L2Norm. This produces the unit vector (vector lenght = 1) pointing to the same direction of the original vector.

## Poor performance of `dotProduct` in clickhouse now

Both `cosineDistance` and `dotProduct` exist in clickhouse. `dotProduct` should be faster, but currently is (~16x) slower .

```sql
WITH cosineDistance(img_emb, <query_emb>) AS score
SELECT id, score, caption
FROM laion
ORDER BY score DESC
LIMIT 3
# 3 rows in set. Elapsed: 0.535 sec. Processed 1.00 million rows, 2.13 GB (1.87 million rows/s., 3.98 GB/s.) Peak memory usage: 164.25 MiB.

WITH dotProduct(img_emb, <query_emb>) AS score
SELECT id, score, caption
FROM laion
ORDER BY score DESC
LIMIT 3
# 3 rows in set. Elapsed: 8.385 sec. Processed 1.00 million rows, 2.13 GB (119.31 thousand rows/s., 254.04 MB/s.) Peak memory usage: 420.67 MiB.
```

## Possible solution

I know projects like [SimSIMD](https://github.com/ashvardanian/SimSIMD) could help with this optimization.

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes #58096

This PR is similar in spirit as #58866.

Review note. The PR includes these commits:
- (please ignore) Refactoring and cleanups: https://github.com/ClickHouse/ClickHouse/commit/0fff496ad388df450af538df2cc3a57c29c86f45
- (please ignore) More refactorings: https://github.com/ClickHouse/ClickHouse/commit/db974098d073496d640bf08f0e43f3aa572ad3ff
- A try to extend the existing tests and make them more systematic: https://github.com/ClickHouse/ClickHouse/commit/877dc695f29b0df163a847ad323eaf2d20552af5
- New performance tests for `dotProduct`: https://github.com/ClickHouse/ClickHouse/commit/7d354164a5c47b762cfac2aea7406cd99fc00fe0
- Some help for the compiler with auto-vectorization (*): https://github.com/ClickHouse/ClickHouse/commit/e57b8f64e0e8112c063daacba4933145000603de
- Manual vectorization for `Float32/Float64` cases (**): https://github.com/ClickHouse/ClickHouse/commit/daa61a8576918ef650e314d06f4ece08cbbc74e7
- (please ignore) Further refactorings to make the code more similar to arrayDistance.cpp: https://github.com/ClickHouse/ClickHouse/pull/60202/commits/9f37f179005089c58f8f7cdacc081e9b203bcf38

(*) For query
   `SELECT sum(dp) FROM (SELECT dotProduct(v, v) AS dp FROM vectors)`
with vectors of type
   `Array(Float32)` and
   `Array(Float64)`
and dimension = 150, better auto-vectorization reduces runtimes from
  0.46 sec to 0.34 sec (`Float32`) and
  0.74 sec to 0.68 sec (`Float64`))

(**) Manual vectorization further improves runtimes from
  0.34 sec to 0.30 sec (`Float32`) and
  0.68 sec to 0.54 sec (`Float64`))

Improvements are not revolutionary but in the same ballpark as in #58866.

@gingerwizard 

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Vectorized function `dotProduct` which is useful for vector search

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit daa61a8576918ef650e314d06f4ece08cbbc74e7 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60202/daa61a8576918ef650e314d06f4ece08cbbc74e7/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60202/daa61a8576918ef650e314d06f4ece08cbbc74e7/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Functions/array/arrayDotProduct.cpp [hunk 1]
```diff
@@ -49,22 +210,114 @@ class ArrayDotProductImpl
...
+#if USE_MULTITARGET_CODE
+        if constexpr ((std::is_same_v<ResultType, Float32> || std::is_same_v<ResultType, Float64>)
+                        && std::is_same_v<ResultType, LeftType> && std::is_same_v<LeftType, RightType>)
+        {
+            if (isArchSupported(TargetArch::AVX512F))
+                accumulateCombine(left, right, size, i, state);
+        }
+#else
+        /// Process chunks in vectorized manner
+        static constexpr size_t VEC_SIZE = 4;
```
Comment (hanfei1991) [inline src/Functions/array/arrayDotProduct.cpp:303]: I’m not familiar with vectorized execution. Just out of curiosity, why 4 is the best number of the loop size?

File: docs/en/sql-reference/functions/tuple-functions.md [hunk 2]
```diff
@@ -542,7 +542,7 @@ Alias: `scalarProduct`.
 
 - Scalar product.
 
-Type: [Int/UInt](../../sql-reference/data-types/int-uint.md), [Float](../../sql-reference/data-types/float.md) or [Decimal](../../sql-reference/data-types/decimal.md).
+Type: [Int/UInt](../../sql-reference/data-types/int-uint.md) or [Float](../../sql-reference/data-types/float.md).
```
Comment (hanfei1991) [inline docs/en/sql-reference/functions/tuple-functions.md:545]: Does it mean we no longer support decimal types?

File: src/Functions/array/arrayDotProduct.cpp (hunk 1 above)
Comment (rschu1ze) [reply src/Functions/array/arrayDotProduct.cpp:303]: In general: Dividing the loop into "sub-loops" helps the compiler to autovectorize better (see e.g. [here](https://llvm.org/docs/Vectorizers.html) for how that works in general). The thing is that - in theory - the compiler could/should chop up a loop by itself, without forcing the programmer to do it manually like here, but for some reason that did not happen for the original code:

```cpp
for (size_t i = 0; i < size; ++i)
    result += static_cast<ResultType>(left[i]) * static_cast<ResultType>(right[i]);
```

About `4`: The choice of the constant is sort of black magic. By default, ClickHouse is compiled with SSE 4.2 instructions as minimum SIMD instruction set (see [here](https://github.com/ClickHouse/ClickHouse/blob/master/cmake/cpu_features.cmake#L89)), which has 128-bit wide SIMD registers. So for input/output types Float32 / Float64 one would expect optimal constant values of 4 / 2. However, based on runtime measurements, the concrete choice (as long as it is a power of 2) does not make any difference in terms of performance ...

So, the only reason I chose `4` was that it was also chosen in similar code in a different function implementation (https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/array/arrayDistance.cpp#L580).

I'll add some more comments.

File: docs/en/sql-reference/functions/tuple-functions.md (hunk 2 above)
Comment (rschu1ze) [reply docs/en/sql-reference/functions/tuple-functions.md:545]: It never did :-)

File: src/Functions/array/arrayDotProduct.cpp (hunk 1 above)
Comment (rschu1ze) [reply src/Functions/array/arrayDotProduct.cpp]: Yeah, that might be!

**Review by hanfei1991 — APPROVED** [review]

**Pull request merged**